component/src/ibc/ibc_handler.rs

```diff
@@ -0,0 +1,189 @@
+    async fn chan_open_init_check(&self, msg: &MsgChannelOpenInit) -> Result<()>;
+    async fn chan_open_try_check(&self, msg: &MsgChannelOpenTry) -> Result<()>;
+    async fn chan_open_ack_check(&self, msg: &MsgChannelOpenAck) -> Result<()>;
+    async fn chan_open_confirm_check(&self, msg: &MsgChannelOpenConfirm) -> Result<()>;
+    async fn chan_close_confirm_check(&self, msg: &MsgChannelCloseConfirm) -> Result<()>;
+    async fn chan_close_init_check(&self, msg: &MsgChannelCloseInit) -> Result<()>;
+
+    async fn recv_packet_check(&self, msg: &MsgRecvPacket) -> Result<()>;
+    async fn timeout_packet_check(&self, msg: &MsgTimeout) -> Result<()>;
+    async fn acknowledge_packet_check(&self, msg: &MsgAcknowledgement) -> Result<()>;
+}
+
+// AppHandlerExecute defines the interface for an IBC application to consume IBC channel and packet
+// events and apply their state transition logic. The IBC component will only call these methods
+// once the transaction has been validated using the AppHandlerCheck interface.
+#[async_trait]
+pub trait AppHandlerExecute: Send + Sync {
+    async fn chan_open_init_execute(&mut self, msg: &MsgChannelOpenInit);
+    async fn chan_open_try_execute(&mut self, msg: &MsgChannelOpenTry);
+    async fn chan_open_ack_execute(&mut self, msg: &MsgChannelOpenAck);
+    async fn chan_open_confirm_execute(&mut self, msg: &MsgChannelOpenConfirm);
+    async fn chan_close_confirm_execute(&mut self, msg: &MsgChannelCloseConfirm);
+    async fn chan_close_init_execute(&mut self, msg: &MsgChannelCloseInit);
+
+    async fn recv_packet_execute(&mut self, msg: &MsgRecvPacket);
+    async fn timeout_packet_execute(&mut self, msg: &MsgTimeout);
+    async fn acknowledge_packet_execute(&mut self, msg: &MsgAcknowledgement);
+}
+
+pub trait AppHandler: AppHandlerCheck + AppHandlerExecute {}
+
+/// an AppRouter is an implementation of AppHandler that is the root router for all IBC
+/// applications. Applications can register themselves on an IBC port by calling AppRouter.bind().
+pub struct AppRouter {
+    handlers: HashMap<PortId, Box<dyn AppHandler>>,
+}
+
+impl AppRouter {
+    pub fn new() -> Self {
+        AppRouter {
+            handlers: HashMap::new(),
+        }
+    }
+
+    /// Bind an IBC application, given by `handler`, to the given port ID. This will panic if there
+    /// is already an application bound to the given port ID.
+    pub fn bind(&mut self, port_id: PortId, handler: Box<dyn AppHandler>) {
+        if self.handlers.contains_key(&port_id) {
+            panic!("AppRouter: handler already bound for port {}", port_id);
+        }
+        self.handlers.insert(port_id, handler);
+    }
+}
+
+#[async_trait]
+impl AppHandlerCheck for AppRouter {
+    async fn chan_open_init_check(&self, msg: &MsgChannelOpenInit) -> Result<()> {
+        if let Some(handler) = self.handlers.get(&msg.port_id) {
+            handler.chan_open_init_check(msg).await?;
+        }
+        Ok(())
```

Comment:
note that the way this is implemented currently means that if we get channel events for a port which we have no applications bound on, then we just no-op. This seems okay, but it might be worth thinking about if we should allow channels to be opened if we have no applications registered for their port